java-checks/src/main/java/org/sonar/java/checks/regex/AbstractRegexCheck.java

```diff
@@ -53,27 +62,84 @@
       .withAnyParameters()
       .build());
 
+  private RegexScannerContext regexContext;
+
   @Override
-  protected MethodMatchers getMethodInvocationMatchers() {
+  public final void setContext(JavaFileScannerContext context) {
+    this.regexContext = (RegexScannerContext) context;
+    super.setContext(context);
+  }
+
+  @Override
+  protected final MethodMatchers getMethodInvocationMatchers() {
     return REGEX_METHODS;
   }
 
   @Override
-  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+  protected final void onMethodInvocationFound(MethodInvocationTree mit) {
     Arguments args = mit.arguments();
     if (args.isEmpty()) {
       return;
     }
-    ExpressionTree arg0 = args.get(0);
-    if (arg0.is(Tree.Kind.STRING_LITERAL)) {
-      checkRegex(((DefaultJavaFileScannerContext) context).regexForLiterals((LiteralTree) arg0), mit);
+    getLiterals(args.get(0))
+      .map(regexContext::regexForLiterals)
+      .ifPresent(result -> checkRegex(result, mit));
+  }
+
+  @VisibleForTesting
+  protected static Optional<LiteralTree[]> getLiterals(ExpressionTree expr) {
+    switch (expr.kind()) {
+      case PLUS:
+        return getLiteralsFromStringConcatenation((BinaryExpressionTree) expr);
+      case IDENTIFIER:
+        return getLiteralsFromFinalVariables((IdentifierTree) expr);
+      case PARENTHESIZED_EXPRESSION:
+        return getLiterals(ExpressionUtils.skipParentheses(expr));
+      case STRING_LITERAL:
+        return Optional.of(new LiteralTree[] {(LiteralTree) expr});
+      default:
+        return Optional.empty();
+    }
+  }
+
+  private static Optional<LiteralTree[]> getLiteralsFromStringConcatenation(BinaryExpressionTree expr) {
+    Optional<LiteralTree[]> leftLiterals = getLiterals(expr.leftOperand());
+    if (!leftLiterals.isPresent()) {
+      return Optional.empty();
+    }
+    Optional<LiteralTree[]> rightLiterals = getLiterals(expr.rightOperand());
+    if (!rightLiterals.isPresent()) {
+      return Optional.empty();
+    }
+    LiteralTree[] combined = Streams.concat(Arrays.stream(leftLiterals.get()), Arrays.stream(rightLiterals.get())).toArray(LiteralTree[]::new);
+    return Optional.of(combined);
+  }
+
+  private static Optional<LiteralTree[]> getLiteralsFromFinalVariables(IdentifierTree identifier) {
+    Symbol symbol = identifier.symbol();
+    if (!symbol.isVariableSymbol()) {
+      return Optional.empty();
+    }
+
+    Symbol.VariableSymbol variableSymbol = (Symbol.VariableSymbol) symbol;
+    if (!variableSymbol.isFinal()) {
```

Comment:
Is there a way to check whether a variable is *effectively* final? Because I'd like this to work with locals and locals are rarely explicitly defined as final.

Reply:
actually, thanks to ECJ, there is! I'm trying this!